portal-kernel/src/com/liferay/portal/kernel/security/auth/PrincipalThreadLocal.java

```diff
@@ -72,11 +82,40 @@ public static void setPassword(String password) {
 		_password.set(password);
 	}
 
+	private static Long _getCompanyId(String name) {
+		if ((name == null) || (GetterUtil.getLong(name) == 0L)) {
+			return CompanyConstants.SYSTEM;
+		}
+
+		Map<String, Long> namesCompanyId = _namesCompanyId.get();
+
+		return namesCompanyId.computeIfAbsent(
+			name,
+			key -> {
+				User user = UserLocalServiceUtil.fetchUser(
+					GetterUtil.getLong(key));
+
+				if (user != null) {
+					return user.getCompanyId();
+				}
+
+				return null;
+			});
+	}
+
 	private static final Log _log = LogFactoryUtil.getLog(
 		PrincipalThreadLocal.class);
 
+	private static final CentralizedThreadLocal<Long> _companyId =
+		new CentralizedThreadLocal<>(
+			PrincipalThreadLocal.class + "._companyId",
+			() -> CompanyConstants.SYSTEM);
 	private static final ThreadLocal<String> _name =
 		new CentralizedThreadLocal<>(PrincipalThreadLocal.class + "._name");
+	private static final CentralizedThreadLocal<Map<String, Long>>
```

Comment:
I can not see the reasons for these 2 new ThreadLocals.

I can see you are just mapping userName -> User -> user's companyId. How is this a threadlocal thing? The lookup chain is always there as long as you still have the userName info which is already a ThreadLocal.

Is this merely meant for caching? How many times we are hitting this cache with same key within the same request scope? Considering the entity is there, I don't think this is worthy at all.

Reply:
Agree, we should probably change this.